### ISSUE TITLE
Support for generic dataclasses

### ISSUE THREAD

**Issue by bilelomrani1** [omni-us/jsonargparse#366]
## 🚀 Feature request

Add support for generic dataclasses in `ArgumentParser`.

### Motivation

I often use generic dataclasses as configuration objects in libraries code. When building a CLI application, I would like to reuse these configurations to build a specialized parser.
Currently if a generic dataclass is used as a subparser, the argument parsing complains that all parameters are not typed and crashes.

```python
from dataclasses import dataclass
from typing import Generic, TypeVar

from jsonargparse import ActionConfigFile, ArgumentParser

T = TypeVar("T")

# This is generic library code
@dataclass
class SubConfig(Generic[T]):
    generic: T
    a: int = 1
    b: float = 2.0

# This is CLI-specific
@dataclass
class Config:
    subconfig: SubConfig[int]


if __name__ == "__main__":
    parser = ArgumentParser(print_config="--print-config")
    parser.add_argument("--load", action=ActionConfigFile, help="Path to a YAML configuration")
    parser.add_dataclass_arguments(Config, "config")
    parser.parse_args()
```

```bash
$ python mre_jsonargparse.py --print-config
Traceback (most recent call last):
  File "/Users/bilelomrani/Documents/ILLUIN.nosync/instructions-finetuning/mre_jsonargparse.py", line 24, in <module>
    parser.add_dataclass_arguments(Config, "config")
  File "/Users/bilelomrani/Documents/ILLUIN.nosync/instructions-finetuning/.venv/lib/python3.10/site-packages/jsonargparse/_signatures.py", line 435, in add_dataclass_arguments
    self._add_signature_parameter(
  File "/Users/bilelomrani/Documents/ILLUIN.nosync/instructions-finetuning/.venv/lib/python3.10/site-packages/jsonargparse/_signatures.py", line 381, in _add_signature_parameter
    raise ValueError(
ValueError: With fail_untyped=True, all mandatory parameters must have a supported type. Parameter 'subconfig' from '__main__.Config.__init__' does not specify a type.
```

With `fail_untyped=False`, the printed config is empty.

### Pitch

Would be cool if `parser.add_dataclass_arguments` would work with generic dataclasses and would generate a specialized subparser (in the case of my example, adding an `int` validator to `config.subconfig.generic`).

**Comment by mauvilsa** [comment]
Support for generic types would be a good addition to jsonargparse.

**Comment by mauvilsa** [comment]
I have implemented support for generic types in #378. @bilelomrani1 please test it.

**Comment by charlie-rivos** [comment]
This doesn't appear to work with `auto_cli`, am I doing something wrong?

Generic version:
```
# generic_example.py
from jsonargparse import auto_cli
from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")

@dataclass
class ExampleGeneric(Generic[T]):
    val: T

@dataclass
class Config:
    example: ExampleGeneric[str]

if __name__ == "__main__":
    config: Config = auto_cli(
        Config,
    )

    print(config)
```


Non generic version:
```
# example.py
from jsonargparse import auto_cli
from dataclasses import dataclass

@dataclass
class ExampleVal:
    val: str

@dataclass
class Config:
    example: ExampleVal

if __name__ == "__main__":
    config: Config = auto_cli(
        Config,
    )

    print(config)
```

Setup:
```
$ python -m venv venv
$ source venv/bin/activate
$ pip install jsonargparse==4.38.0
```

Basic example:
```
$ python example.py --example.val 'val'
Config(example=ExampleVal(val='val'))
```

Generic example:
```
$ python generic_example.py --example.val 'val' 
Traceback (most recent call last):
  File "/home/charlie/example/generic_example.py", line 16, in <module>
    config: Config = auto_cli(
  File "/home/charlie/example/venv/lib/python3.10/site-packages/jsonargparse/_cli.py", line 102, in auto_cli
    return _run_component(components, init)
  File "/home/charlie/example/venv/lib/python3.10/site-packages/jsonargparse/_cli.py", line 210, in _run_component
    return component(**cfg)
TypeError: Config.__init__() got an unexpected keyword argument 'example.val'
```

**Comment by charlie-rivos** [comment]
I should have posted this instead. I thought I had narrowed down my problem with the above that may be an unrelated issue. My application has multiple layers of generics and jsonargparse doesn't seem to support this any longer.

```
from jsonargparse import CLI
from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")

@dataclass(frozen=True, kw_only=True)
class ExampleGeneric2(Generic[T]):
    value: T

@dataclass(frozen=True, kw_only=True)
class ExampleGeneric(Generic[T]):
    workloads: tuple[ExampleGeneric2[T], ...]

@dataclass(frozen=True, kw_only=True)
class Config(ExampleGeneric[str]):
    workloads: tuple[ExampleGeneric2[str], ...]

if __name__ == "__main__":
    config: Config = CLI(
        Config,
    )

    print(config)
```

I am trying to uprev the jsonargparse version used by my project. On `4.38.0` I get the following error:
```
$ python generic_example.py '[{value: str}]'
usage: generic_example.py [-h] [--config CONFIG] [--print_config[=flags]] [ITEM,...]
error: Parser key "workloads":
  Invalid or unsupported input: class=<class 'typing._Final'>, method_or_property=__call__
```


On jsonargparse 4.27.1:
```
$ python generic_example.py '[{value: str}]'
Config(workloads=(ExampleGeneric2(value='str'),))
```

**Comment by mauvilsa** [comment]
@charlie-rivos thank you for reporting. This should now be fixed with #709. Though, next time if you find a bug or a regression, please create a new issue.

**Comment by charlie-rivos** [comment]
Will do, thank you for your fixes!
